proposals/lambdas-04-kinds.md

```diff
@@ -0,0 +1,203 @@
+  - can receive non-sendable yielded values
+  - can return a non-sendable yield result value
+
+Mobility:
+  - CANNOT cross an actor boundary
+  - CANNOT be objectified or stored away for later (yields must occur in the immediate scope of the called function)
+
+Theoretical "as if" modeling:
+  - It's as if the called function state were kept on the stack to pause and continue execution across multiple intermediate returns
+
+Internal Implementation / Optimization:
+  - the block has no reference capability to track, as it is not objectified
+  - zero or more calls with intermediate return values representing yields from the called function
+  - stack-allocated "continuation struct" persists stack state for called function across multiple calls
+  - for a self-recursive yielding function, the continuation struct must be heap-allocated instead of stack-allocated, but the rest of the semantics are not affected.
+
+## Kind B (single-use isolated lambda)
+
+Call count:
+  - can be called zero or one time
+  - CANNOT be called more than once
+
+Inter-scope interaction:
+  - can capture variables from the outer scope
+  - can rebind captured variables in the inner scope (but not the outer scope)
+  - can consume `iso` variables in the inner scope, provided that they aren't used any more in the outer scope (and thus the consume does not affect the outer scope)
+  - CANNOT affect variable bindings in the outer scope
+  - CANNOT raise an error into the outer scope
+  - CANNOT return early, exiting the outer scope
+
+Ref cap restrictions:
+  - CANNOT capture non-sendable values from the outer scope
+  - can receive non-sendable yielded values
+  - can return a non-sendable yield result value
+
+Mobility:
+  - can cross an actor boundary
+  - can be objectified and stored away for later, but only as an `iso` with `:fun iso` so as to be single-use
+
+Theoretical "as if" modeling:
+  - It's as if the block becomes an object (with ref cap: `iso`)
+  - the object has a single function corresponding to the block body
+  - the captured variables are arguments to a `:new iso` constructor
+  - within the block, the captured variables are fields, which may be rebound internally
+  - the block function is a `:fun iso` that immediately lowers itself to `ref`, so that the captured variable fields may be seen through the un-colored `ref` lens.
+  - the `ref` is not recovered to an `iso`, so the function is not callable again
+
+Internal Implementation / Optimization:
+  - when objectified, it is a heap-allocated object of an anonymous one-off type, behaving like the "as if" model, and using a virtual table for dispatch when called.
+  - in some cases, virtual dispatch on the object may be possible to avoid by using automatic specialization optimizations in the late compiler stages.
+  - a second block that is tail-chained to it without accruing additional captures may be optimized to be another virtual method attached to the same runtime object
```

Comment:
When talking about "tail-chained" blocks here I am referring to the aforementioned concept that was explained in the bank account transaction example, wherein a block section would end with another `...->` symbol followed by a new section in the same block. Basically, syntax that looks like this:

```savi
@foo ...> (
| x |
  # do stuff with x
  @bar ...->
| y |
  # do stuff with y
)
```

Which is syntax sugar for:
  
```savi
@foo ...> (x |
  # do stuff with x
  @bar ...-> (y |
    # do stuff with y
  )
)
```

But when we recognize either form as a tail-chain with the second async lambda's execution following the first async lambda's execution, then as long as certain conditions are met we can quietly combine them into the same heap-allocated object to avoid an extra allocation.

> Won't this need to be a new virtual table anyway?

This optimization will only be possible when the potentially-combined lambdas have different virtual table index numbers, so that they will not collide if occupied in the same table. This should always be possible in cases where the lambda signatures (their params and return type) are different/incompatible, but may not be possible in cases where the signatures are the same/compatible.